src/plugins/curator/components/gallery-media/index.jsx

```diff
@@ -79,58 +79,52 @@ class GalleryMedia extends React.Component {
     this.handleSlideImagesPerRow = this.handleSlideImagesPerRow.bind(this);
+    this.handleSubmitReorder = this.handleSubmitReorder.bind(this);
     this.handleToggleCardOverlay = this.handleToggleCardOverlay.bind(this);
     this.handleToggleModal = this.handleToggleModal.bind(this);
   }
 
   componentDidMount() {
-    const { delegate } = this.props;
+    const { delegate, history } = this.props;
     delegate.handleSearchGalleryAssets();
+    this.unblock = history.block((location, action) => {
+      const { reorder } = this.state;
+      const { nodes } = this.props;
+      const movedNodes = pickBy(
+        reorder.nodesToUpdate || {},
+        (sequence, id) => nodes.findIndex((node) => node.get('_id').toString() === id)
+          !== reorder.nodes.findIndex((node) => node.get('_id').toString() === id),
+      );
+      if (Object.keys(movedNodes).length) {
+        return 'You have unsaved changes. Are you sure you want to leave?';
+      }
+      return true;
+    });
   }
 
-  /**
-   * check to make sure the nodes we get back from the search request have been set as expected and
-   * if not, search again. necessary because sometimes we get notified that the reorder operation
-   * is done but since the search is performed so quickly after the reorder "finishes", it may
-   * bring in older versions of the assets.
-   */
-  componentDidUpdate({ items: prevItems }) {
-    const { delegate, items } = this.props;
-    const { reorder: { isWaiting, itemsToUpdate } } = this.state;
-    if (isWaiting && !isEqual(prevItems, items)) {
-      let isValid = true;
-      const keys = Object.keys(itemsToUpdate);
-      for (let i = 0; i < keys.length; i += 1) {
-        const item = items.find((itm) => itm.assetId.toString() === keys[i]);
-        if (item.gallerySequence !== itemsToUpdate[keys[i]]) {
-          isValid = false;
-          break;
-        }
-      }
-      if (!isValid) {
-        toast.show();
-        delegate.handleSearchGalleryAssets();
-      } else {
-        toast.close();
-        this.setState(() => ({ // eslint-disable-line react/no-did-update-set-state
-          reorder: {
-            isWaiting: false,
-            itemsToUpdate: [],
-          },
-        }));
-      }
+  componentDidUpdate({ nodes: prevNodes }) {
+    const { nodes } = this.props;
+    const { reorder: { nodesToUpdate } } = this.state;
+    if ((prevNodes.length !== nodes.length) && nodesToUpdate) {
+      this.handleReorderOnGalleryChanged();
     }
   }
 
   componentWillUnmount() {
+    this.unblock();
     const { delegate } = this.props;
     delegate.handleClearChannel();
   }
 
   async handleAddAssets(assetMap) {
     const { delegate } = this.props;
-    await delegate.handleAddGalleryAssets(assetMap);
+    try {
+      await delegate.handleAddGalleryAssets(assetMap);
+    } catch (e) {
+      // did not update, should restore tab
+    }
     delegate.handleSearchGalleryAssets();
 
     return Promise.resolve();
```

Comment:
@hunyoboy an async function already returns a promise.  you don't need to make one yourself